### PR TITLE
System test for Preserve return_to param when switching from /login to /signup

### LINKED ISSUE #5585: System test for Preserve `return_to` param when switching from /login to /signup
Write System test for Preserve `return_to` param when switching from /login to /signup

### PR DESCRIPTION
**Pull request by SidharthBansal**
Fixes #5585

### CODE REVIEW

**Comment by SidharthBansal** [comment]
@jywarren another PR completed. @gauravano @sashadev-sky @jywarren please review and suggest changes if any.
Thanks

File: test/system/users_test.rb [hunk 1]
```diff
@@ -0,0 +1,16 @@
+require "application_system_test_case"
+
+class UsersTest < ApplicationSystemTestCase
+  # test "visiting the index" do
+  #   visit users_url
+  #
+  #   assert_selector "h1", text: "Users"
+  # end
```
Comment (siaw23) [inline test/system/users_test.rb]: Lets' remove commented out code.

File: test/system/users_test.rb [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+  test "preserve return_to while switching from login to signup" do
+    visit '/login?return_to=/subscribe/multiple/tag/tag1,tag2'
+    click_on "sign up"
+    assert_text "Create a password"
+    assert_text "Confirmation"
+    assert_text "Bio"
```
Comment (siaw23) [inline test/system/users_test.rb]: I don't think this test does what you mean or what I understand from the title. You want to test that a user is returned to a URL after they try to login and it appears they don't have an account, so after they sign up, they are returned to the URL they were trying to access?

Comment (SidharthBansal) [reply test/system/users_test.rb]: I think you have not understood the functionality properly. This is the continued work of #5559.
The test ensures that when the user clicks on the ``sign up `` button on the login page with url `/login?return_to=x` ,then they get redirected to the signup page with the url '/signup?return_to=x`.

I am sorry for the inconvenience.

Comment (SidharthBansal) [reply test/system/users_test.rb]: Thanks for the input. I have modified the tests according to your suggestions.

File: test/system/users_test.rb [hunk 3]
```diff
@@ -0,0 +1,12 @@
+require "application_system_test_case"
+
+class UsersTest < ApplicationSystemTestCase
+  test "preserve return_to while switching from login to signup" do
+    visit '/login?return_to=/subscribe/multiple/tag/tag1,tag2'
+    click_on "sign up"
+    assert_redirected_to '/signup?return_to=/subscribe/multiple/tag/tag1,tag2'
```
Comment (SidharthBansal) [inline test/system/users_test.rb]: @siaw23 @gauravano @jywarren can you please help me why this line shows error?

Comment (SidharthBansal) [reply test/system/users_test.rb]: line 7 i mean

Comment (siaw23) [reply test/system/users_test.rb]: Line 7 is an assertion. Which error do you mean @SidharthBansal? Can you please paste it here?

**Comment by SidharthBansal** [comment]
Please see the travis
It is failing

File: test/system/users_test.rb (hunk 3 above)
Comment (siaw23) [reply test/system/users_test.rb]: Right. So you don't do assertion here like this. When you `click_on "sign up"` you have to assert the presence of some text on that page after clicking. So try manually clicking on "sign up" then assert the presence of some text there like so `assert_text 'Some Text'` where "Some Text" is some unique text that you can only see after clicking "sign up". Let's try that if that works.

**Comment by SidharthBansal** [comment]
Hi, I have already written statements that confirm the presence of text
which appears on the signup page
I need to confirm that I am routed to
/signup?return_to=subscribe/multiple/tag/tag1,tag2 page
How shall I do it?
Any help much appreciated.

**Comment by siaw23** [comment]
Now I see the assertion `assert_text "Create a password`. That's enough. That assertion passing is your confirmation that you've been redirected. To test you're redirected again back to the initial page you visited, something has to happen right? So on the "creating password" page another button has to be clicked, THEN you can make a new text assertion. Don't use `assert_redirected_to` here. Click buttons and assert the presence of texts.

**Comment by SidharthBansal** [comment]
Let me try this. Thanks

File: test/system/users_test.rb (hunk 3 above)
Comment (SidharthBansal) [reply test/system/users_test.rb]: Test will fail as we don't have access to Recaptcha feature in the testing environment.
So, we will not be able to submit the form by clicking on the  "Sign Up" button.

Comment (SidharthBansal) [reply test/system/users_test.rb]: Can we try unit test for this? We just want three steps.
1. User clicks on '/login?return_to=X'
2. User clicks on "Signup"
3. User is now at '/signup?return_to=X'
But I don't know how to do statement 2 in unit tests.

**Comment by jywarren** [comment]
Noting that we may have some similarities in debugging in https://github.com/publiclab/plots2/pull/5605 -- 🙌 

System tests are hard to get used to but they're coming along!!!

**Comment by SidharthBansal** [comment]
back on it after a long time

**Pull request closed without merging**